Prevent excessive api calls

ghubstan opened the pull request:
This change provides a gRPC `CallRateMeteringInterceptor` to help protect the server and network against being overloaded by CLI scripting mistakes.

An interceptor instance can be configured on a gRPC service to set per method call rate limits on one or more of a service's methods. For example, the `GrpcOffersService` could be configured with this interceptor to set the `createoffer` rate limit to 5/hour, and the `takeoffer` call rate limit could be set to 20/day.  Whenever a call rate limit is exceeded, the gRPC call is aborted and the client receives a "rate limit exceeded" error.

Below is a simple example showing how to set rate limits for only the `getVersion` method in `GrpcVersionService`.

    final ServerInterceptor[] interceptors() {
        return new ServerInterceptor[]{
                new CallRateMeteringInterceptor(new HashMap<>() {{
                    put("getVersion", new GrpcCallRateMeter(2, SECONDS));
                }})
        };
    }

It specifies a CLI can execute `getversion` 2 times / second.

This is not a throttling mechanism, there is no blocking nor locking to slow call rates.  When call rates are exceeded, calls are simply aborted.

This is the 10th in a chain of PRs beginning with https://github.com/bisq-network/bisq/pull/4884.
PR https://github.com/bisq-network/bisq/pull/4960 should be reviewed before this one.

ghubstan committed:
Refactor BtcWalletService to let api override fee rates

BtcWalletService was changed to allow the api to override tx fee
rates from the sendbsq and sendbtc methods.  The api methods will
still be able to use the network fee service and custom tx fee rate
preference, and set / unset the custom tx fee rate preference, but
the change will permit the addition of an optional txFeeRate parameter
to the sendbsq and sendbtc methods (todo).  A few other minor changes
(style and removal of never thrown ex spec) were also made to this class.

Two BtcWalletService methods were refactored.

- The redundant (was always true) boolean isSendTx argument was removed
  from the completePreparedVoteRevealTx method signature.

- The redundant (was always true) boolean useCustomTxFee was removed
  from the completePreparedBsqTx method signature.

- The completePreparedSendBsqTx method was overloaded with a 2nd parameter
  (Coin txFeePerVbyte) to allow api to override fee service and custom
  tx fee rate when sending BSQ or BTC.

- The completePreparedBsqTx method was overloaded with a 3rd parameter
  (Coin txFeePerVbyte) to allow api to override fee service and custom
  tx fee rate when sending BSQ or BTC.

The following line was deleted from the completePreparedBsqTx method
because txFeePerVbyte is now an argument:

	Coin txFeePerVbyte = useCustomTxFee ? getTxFeeForWithdrawalPerVbyte() : feeService.getTxFeePerVbyte();

This useCustomTxFee value was always true, and redudant here because
getTxFeeForWithdrawalPerVbyte() returns feeService.getTxFeePerVbyte()
or the custom fee rate preference. i.e.,

Coin txFeePerVbyte = useCustomTxFee ? getTxFeeForWithdrawalPerVbyte() : feeService.getTxFeePerVbyte();

	is equivalent to

Coin txFeePerVbyte = getTxFeeForWithdrawalPerVbyte();

LockupTxService, UnlockTxService, BsqSendView, and BsqTransferService
were adjusted to this BtcWalletService refactoring.

ghubstan committed:
Merge branch 'master' into 02-refactor-completePreparedSendBsqTx

ghubstan committed:
Add optional txFeeRate parameter to api sendbsq

If present in the sendbsq command, the parameter will override the fee
service and custom fee rate setting for the BSQ transaction.

Also changed the sendbsq grpc return type to a lightweight TX proto wrapper.

Besides some small refactoring in the CLI, all the changes are
adjustments for this new sendbsq parameter and its new grpc return value.

ghubstan committed:
Merge branch 'master' into 03-add-txFeeRate-param

ghubstan committed:
Add new api method 'sendbtc' and test

Takes an address, amount, and optional txfeerate param,
returns a lightweight TxInfo proto.

Also overloaded two BtcWalletService methods to allow sendbtc
to pass in the tx fee rate -- overriding the fee service and
custom fee rate setting.

ghubstan committed:
Merge branch 'master' into 04-add-sendbtc-impl

ghubstan committed:
Support tx memo field for btc withdrawals from api

- Added optional memo parameter to the api's sendbtc and
  withdrawfunds commands.

- Removed the @Nullable annotation was removed because protobuf
  does not support null.

- Visibility in two wallet check methods were changed from private
  to pkg protected so the CoreTradeService could use them.

- Adjusted affected tests.  (Asserting the memo field was set on a
  transaction cannot be checked from apitest yet.)

ghubstan committed:
Remove unused imports

ghubstan committed:
Merge branch 'master' into 05-use-memo-tx-field

ghubstan committed:
Use Bisq's UserThread.executor in gRPC server

ghubstan committed:
Append nullable withdrawalTxId field to Trade proto message

The withdrawalTxId field will be set in TradeManager#onWithdrawRequest
upon successful trade completion and withdrawal of funds.

Persisting withdrawalTxId will allow the api and ui to find the withdrawalTxId
for a completed trade after the seller withdraws funds to an external wallet.
In turn, the withdrawal tx's memo field will be accessible in a new (todo)
api getTx(txID) api method.

Changed:

- Appended field 'string withdrawal_tx_id = 40' to pb.proto's Trade message.

- Added nullable 'String withdrawalTxId' to Trade entity class.

- Added trade.setWithdrawalTxId(transaction.getTxId().toString()) in
  TradeManager#onWithdrawRequest's callback.

ghubstan committed:
Add new api method gettransaction

This change was prompted by the recent changes in the main branch to
allow a tx memo field to be set from the UI and API.

This and the prior PR address the API's need to be able to fetch a
tx (with a memo).  The API can now get a completed trade's withdrawal
txid and pass it as a gettransaction parameter.

See previous PR "Append nullable withdrawalTxId field to Trade".

	https://github.com/bisq-network/bisq/pull/4937

A summary of changes by file:

grpc.proto

- Added withdrawalTxId field to existing TradeInfo proto & wrapper.
- Reordered fields in TradeInfo proto.
- Added new fields to be displayed by TxInfo proto in CLI.
- Fixed typo: unsetTxFeeRatePreference -> UnsetTxFeeRatePreference.
- Added new GetTransaction rpc.

GrpcWalletsService - Added new getTransaction gRPC boilerplate.

CoreWalletsService - Added new getTransaction implementation.

TxInfo - Added the new fields for displaying a tx summary from CLI.
This is not intended to be more than a brief summary;  a block explorer
or bitcoin-core client should be used to see the complete definition.

TradeInfo - Added the new withdrawalTxId field defined in grpc.proto.

CliMain - Added new 'case gettransaction'.

TransactionFormat - Formats a TxInfo sent from the server to CLI.

ColumnHeaderConstants - Added console headers used by TransactionFormat.

TradeFormat - Displays a completed trade's WithdrawalTxId if present.

Apitest - Adjusted affected tests: assert tx memo is persisted and
test gettransaction.

ghubstan committed:
Merge branch 'master' into 08-scratch

ghubstan committed:
Adjust create TransferwiseAccount test

As per commit 88f26f93241af698ae689bf081205d0f9dc929fa,
do not autofill all currencies by default but keep all unselected.

ghubstan committed:
Disable method test to avoid repetition

ghubstan committed:
Merge branch 'master' into 09-scratch

ghubstan committed:
Add new CoreApiExceptionHandler to gRPC services

This change reduces gRPC service error handling duplication by moving
it into a @Singleton encapsulating everything needed to wrap
an expected or unexpected core api exception into a gRPC
StatusRuntimeException before sending it to the client.  It also
fixes some boilerpate classes were gRPC error handling was missing.

ghubstan committed:
Merge branch 'master' into 09-refactor-grpc-error-handling

ghubstan committed:
Fix class level comment

ghubstan committed:
Rename gRPC exception handler class

ghubstan committed:
Create grpc interceptor pkg, move auth interceptor into it

ghubstan committed:
Put arguments on separate lines

ghubstan committed:
Prevent excessive api calls

This change provides a gRPC CallRateMeteringInterceptor to help protect
the server and network against being overloaded by CLI scripting mistakes.

An interceptor instance can be configured on a gRPC service to set
individual method call rate limits on one or more of the the service's
methods. For example, the GrpcOffersService could be configured with
this interceptor to set the createoffer rate limit to 5/hour, and
the takeoffer call rate limit could be set to 20/day.  Whenever a
call rate limit is exceeded, the gRPC call is aborted and the client
recieves a "rate limit exceeded" error.

Below is a simple example showing how to set rate limits for one method
in GrpcVersionService.

    final ServerInterceptor[] interceptors() {
        return new ServerInterceptor[]{
                new CallRateMeteringInterceptor(new HashMap<>() {{
                    put("getVersion", new GrpcCallRateMeter(2, SECONDS));
                }})
        };
    }

It specifies a CLI can execute getversion 2 times / second.

This is not a throttling mechansim, there is no blocking nor locking
to slow call rates.  When call rates are exceeded, calls are
simply aborted.

ghubstan committed:
Change long to int, tidy up error msg

ghubstan committed:
Add license comment, stop & toString methods, and make isRunning transient

ghubstan committed:
Make CallRateMeteringInterceptor configurable via json

This adds a GrpcServiceRateMeteringConfig class that can read and
write rate metering interceptor config files, and configure
a gRPC rate metering service interceptor at startup.

This seems excessive, but we need to be able to test and tune
method rate metering without having to change hard coded, default
interceptor rate meters.

ghubstan committed:
Add unit test dependencies to daemon subproject

ghubstan committed:
Add GrpcServiceRateMeteringConfigTest

ghubstan committed:
Add license note

ghubstan committed:
Add license note and toString method

ghubstan committed:
Inject Config into CoreApi

ghubstan committed:
Don't cancel gRPC call if an interceptor does not meter all methods

ghubstan committed:
Configure GrpcVersionService's rate metering interceptor

This change demonstrates how a method call rate metering interceptor
is configured for a gRPC service.  GrpcVersionService uses a custom
rate metering interceptor, or none.  A commented out, 'default'
interceptor is defined as a usage example.

ghubstan committed:
Add ApiTestConfig option --callRateMeteringConfigPath

Points to a call rate metering interceptor configuration file.

Test cases can build a config file, and the test harness will
install it into a daemon's appDataDir before startup.
The installed config file will be used to configure gRPC
service rate metering interceptors.

ghubstan committed:
Install call rate metering config file before startup

Copy the config file at --callRateMeteringConfigPath to each daemon's
appDataDir, where it will be detected at server startup.

ghubstan committed:
Refactor testcase superclasses to support rate metering configs

ghubstan committed:
Test CallRateMeteringInterceptor

ghubstan committed:
Remove unused import

ghubstan committed:
Revert "Append nullable withdrawalTxId field to Trade proto message"

This reverts commit 6aa385e494eb8fa870257c76e078108607503d03.

ghubstan committed:
Ajust TradeInfo to reverting 6aa385e494eb8fa870257c76e078108607503d03

ghubstan committed:
Adjust grpc.proto to reverting 6aa385e494eb8fa870257c76e078108607503d03

ghubstan committed:
Adjust TradeFormat to reverting 6aa385e494eb8fa870257c76e078108607503d03

ghubstan committed:
Adjust TakeSellBTCOfferTest to reverting 6aa385e494eb8fa870257c76e078108607503d03

ghubstan committed:
Adjust TradeTest to reverting 6aa385e494eb8fa870257c76e078108607503d03

ghubstan committed:
Resolve file conflict w/ master

ghubstan committed:
Merge branch '08-gettransaction' into 09-refactor-grpc-error-handling

Adjust to reverting 6aa385e.

ghubstan committed:
Merge branch '09-refactor-grpc-error-handling' into 10-callrate-interceptor

Adjust to reverting to reverting 6aa385e, and fix test file conflict.

sqrrm commented on daemon/src/main/java/bisq/daemon/grpc/interceptor/CallRateMeteringInterceptor.java:
```diff
@@ -0,0 +1,107 @@
+                                                         ServerCall<?, ?> serverCall)
+            throws StatusRuntimeException {
+        String msg = getDefaultRateExceededError(methodName, rateMeter);
+        log.error(StringUtils.capitalize(msg) + ".");
```
This should probably be a `log.warn` as it's not an error but rather expected when the rate limit is exceeded. Might consider changing the method name as well for the same reason.

ghubstan replied:
Fixed in commit 01546ad.

sqrrm commented on daemon/src/main/java/bisq/daemon/grpc/interceptor/GrpcCallRateMeter.java:
```diff
@@ -0,0 +1,65 @@
+    @Getter
+    private final long allowedCallsPerTimeUnit;
+    @Getter
+    private final TimeUnit timeUnit;
```
I find this a bit odd. I think it makes more sense to set the time window in seconds and just have a `int` with number of seconds in the time window.

ghubstan replied:
Fixed in commit 01546ad.  The rate meter can still use a DAYS and HOURS TimeUnit, though I supposed most method rate meters will restrict calls on a per second basis.

sqrrm commented on daemon/src/main/java/bisq/daemon/grpc/interceptor/GrpcCallRateMeter.java:
```diff
@@ -0,0 +1,65 @@
+        if (timer != null)
+            timer.stop();
+
+        timer = UserThread.runPeriodically(() -> callsCount = 0, 1, timeUnit);
```
My expectation is to have a moving time window with the rate limit. The way it's done now I can send double the max amount if I send all at the end of the first period and then continues at the beginning of the second period.

ghubstan replied:
I rewrote the rate meter class in commit 01546ad.  There is a running call count over the current time window, instead of using this timer (oops).

sqrrm commented on daemon/src/main/java/bisq/daemon/grpc/interceptor/GrpcServiceRateMeteringConfig.java:
```diff
@@ -0,0 +1,268 @@
+
+        public File build() {
+            File tmpFile = serializeRateMeterDefinitions();
+            File configFile = new File("/tmp/ratemeters.json");
```
Not sure this is right unless it's only supposed to work on *nix.

ghubstan replied:
It is only supposed to work on *nix but I changed line 233 to
`File configFile = Paths.get(getProperty("java.io.tmpdir"), "ratemeters.json").toFile();`
(Commit 01546ad)

sqrrm reviewed (changes requested):
Mainly I would like to see a moving time window for the rate limit.

Also, I might have missed but is the api only supposed to work on *nix systems? Seems very specific in many places.

ghubstan committed:
Use a simpler, time windowing call rate meter

Rewrote the GrpcCallRateMeter class and adjusted afected classes.

These changes were requested in PR review
https://github.com/bisq-network/bisq/pull/4966#pullrequestreview-557040093

ghubstan committed:
Remove unused local var

ghubstan committed:
Remove redundant callCount field

The size of the timestamp queue is the call count

ghubstan committed:
Fix comment

ghubstan commented:
Force travis jar download.

ghubstan committed:
Support more fine grained rate metering

We need to be able to define call rate meters for time spans not limited
to TimeUnit intervals of 1 SECOND, 1 HOUR, or 1 DAY.  This change allows
more flexibility, e.g., 10 per 5 seconds, 10 per 5 hrs, 100 per 7 days.

ghubstan committed:
Disable CallRateMeteringInterceptorTest and run it from test suite

This will reduce the entire apitest suite's exec time

sqrrm commented on daemon/src/main/java/bisq/daemon/grpc/interceptor/GrpcCallRateMeter.java:
```diff
@@ -1,95 +1,92 @@
-    private Timer timer;
+    private transient final ArrayDeque<Long> callTimestamps;
 
     public GrpcCallRateMeter(int allowedCallsPerTimeUnit, TimeUnit timeUnit) {
```
`allowedCallsPerTimeUnit` is a bit misleading as it's really allowedCallsPerTimeWindow.

ghubstan replied:
Fixed in commit 10727fc0830330e004351cd0567b994173a5d5c5.

sqrrm commented on daemon/src/main/java/bisq/daemon/grpc/interceptor/GrpcCallRateMeter.java:
```diff
@@ -1,95 +1,92 @@
 
-    public void incrementCallsCount() {
-        callsCount++;
+    public boolean isAllowed() {
```
`isAllowed()` indicates it's just checking if something is allowed, this method i changing the state of the class. I would prefer something `call(successCallback, failCallback)` or `checkAndIncrement()`

ghubstan replied:
Fixed in commit 10727fc0830330e004351cd0567b994173a5d5c5.

sqrrm reviewed (changes requested):
Functionality now looks right. Just some comments on naming.

ghubstan committed:
Fix GrpcCallRateMeter method and variable name

- Change method isAllowed() -> checkAndIncrement().

- Change variable allowedCallsPerTimeUnit -> allowedCallsPerTimeWindow.

sqrrm reviewed (approved):
utACK
